tensorpack/tfutils/export.py

```diff
@@ -1,71 +1,85 @@
-                e.export('train_log/mnist_superresolution/checkpoint', 'export/first_export')
-
-            Will generate a model for TensorFlow serving with input 'lowres' and
-            output 'prediction'. The model is in the directory 'export' and can be
-            loaded by
-
-            .. code-block:: python
-
-                import tensorflow as tf
-                from tensorflow.python.saved_model import tag_constants
-
-                export_dir = 'export/first_export'
-                with tf.Session(graph=tf.Graph(), config=tf.ConfigProto(allow_soft_placement=True)) as sess:
-                    tf.saved_model.loader.load(sess, [tag_constants.SERVING], export_dir)
-
-                    prediction = tf.get_default_graph().get_tensor_by_name('prediction:0')
-                    lowres = tf.get_default_graph().get_tensor_by_name('lowres:0')
-
-                    prediction = sess.run(prediction, {lowres: ...})[0]
-
         Args:
-            model (ModelDescBase): the model description which should be exported
-            input_names (list(str)): names of input tensors
-            output_names (list(str)): names of output tensors
+            config (PredictConfig): the config to use.
         """
+        super(ModelExporter, self).__init__()
+        self.config = config
 
-        assert isinstance(input_names, list)
-        assert isinstance(output_names, list)
-        assert isinstance(model, ModelDescBase)
-        self.model = model
-        self.output_names = output_names
-        self.input_names = input_names
+    def export_compact(self, filename, dtypes=None):
+        """Create a self-contained inference-only graph and write final graph to disk.
 
-    def export(self, checkpoint, export_path,
-               tags=[tf.saved_model.tag_constants.SERVING],
-               signature_name='prediction_pipeline'):
+        Args:
+            export_graph_file (str): path to final local of the graph
+            dtype (TYPE, optional): The placeholder data type, or
+                a list that specifies one value per input node name.
         """
+        self.graph = self.config._maybe_create_graph()
+        with self.graph.as_default():
+            input = PlaceholderInput()
+            input.setup(self.config.inputs_desc)
+            with PredictTowerContext(''):
+                self.config.tower_func(*input.get_input_tensors())
+
+            input_tensors = get_tensors_by_names(self.config.input_names)
+            output_tensors = get_tensors_by_names(self.config.output_names)
+
+            self.config.session_init._setup_graph()
+            # we cannot use "self.config.session_creator.create_session()" here since it finalizes the graph
+            sess = tf.Session(config=tf.ConfigProto(allow_soft_placement=True))
+            self.config.session_init._run_init(sess)
+
+            if dtypes is None:
+                dtypes = [n.dtype for n in input_tensors]
```

Comment:
Is there a reasonable use case for this `dtypes` option? (i.e. to not use the default?)

Reply:
For images `InferenceModel` the `dtype` here needs to be `tf.uint8` although the placeholder has type `tf.string`. There are some strange corner cases, which I had.

Reply:
What is the issue? If the dtype parameter for `optimize_for_inference_lib.optimize_for_inference` cannot match `tensor.dtype`, that sounds like a tensorflow bug to me, is it?

Reply:
I cannot reproduce it as:

```python
import tensorflow as tf
from tensorflow.python.tools import optimize_for_inference_lib
from tensorflow.python.framework import graph_util
from tensorflow.python.platform import gfile

p = tf.placeholder(tf.string, (None), name='input_img_bytes')
x = tf.image.decode_png(p, channels=3)
x = tf.expand_dims(x, axis=0)

unused = tf.cast(x, tf.float32)
unused = tf.layers.conv2d(unused, 3, name='conv', kernel_size=[1, 1])

x = tf.squeeze(x, axis=0)
x = tf.image.encode_png(x)
o = tf.identity(x, name='final_bytes')

with tf.Session() as sess:
  sess.run(tf.global_variables_initializer())
  graph_def = sess.graph.as_graph_def()

  graph_def = graph_util.convert_variables_to_constants(
      sess, graph_def, ['final_bytes'])

  the_dtype = p.dtype.as_datatype_enum
  graph_def = optimize_for_inference_lib.optimize_for_inference(graph_def,
                                                                ['input_img_bytes'], ['final_bytes'], [the_dtype], False)

  with gfile.FastGFile('/tmp/tmp_graph.pb', "wb") as f:
    f.write(graph_def.SerializeToString())

tf.reset_default_graph()

# ---------inference ------------------------

with tf.Session(config=tf.ConfigProto(allow_soft_placement=True)) as sess:
  with tf.gfile.GFile('/tmp/tmp_graph.pb', "rb") as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())
    tf.import_graph_def(graph_def)

  print([n.name for n in sess.graph.as_graph_def().node])

  input_img = sess.graph.get_tensor_by_name('import/input_img_bytes:0')
  prediction_img = sess.graph.get_tensor_by_name('import/final_bytes:0')

  with open('lena.png', 'rb') as f:
    buf = bytes(f.read())

  prediction = sess.run(prediction_img, {input_img: buf})

  with open('lena2.png', 'wb') as f:
    f.write(prediction)
```

works fairly well. However, I had some troubles. I am fine with removing the argument. The issues is probably related to `tf.map_fn`.

Reply:
I got error when froze mask rcnn with optimize_for_inference_lib.optimize_for_inference:

File "inference.py", line 35, in <module>
    tf.import_graph_def(od_graph_def)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/importer.py", line 422, in import_graph_def
    raise ValueError(str(e))
ValueError: NodeDef expected inputs '' do not match 1 inputs specified; Op<name=Const; signature= -> output:dtype; attr=value:tensor; attr=dtype:type>; NodeDef: {{node import/output/map/while/add/y}} = Const[dtype=DT_INT32, value=Tensor<type: int32 shape: [] values: 1>](import/output/map/while/Switch:1)

Reply:
For anyone to better understand and diagnose your issue, please post relevant details following the issue template (click "New Issue" -> "Unexpected Problems / Bugs", or visit [this link to post issues about unexpected problems](https://github.com/tensorpack/tensorpack/issues/new?template=unexpected-problems---bugs.md))